controllers/helmreleaseproxy/helmreleaseproxy_controller.go

```diff
@@ -51,11 +54,28 @@ type HelmReleaseProxyReconciler struct {
 func (r *HelmReleaseProxyReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, options controller.Options) error {
 	log := ctrl.LoggerFrom(ctx)
 
+	clusterToHelmReleaseProxies, err := util.ClusterToTypedObjectsMapper(mgr.GetClient(), &addonsv1alpha1.HelmReleaseProxyList{}, mgr.GetScheme())
+	if err != nil {
+		return err
+	}
+
 	return ctrl.NewControllerManagedBy(mgr).
 		WithOptions(options).
 		For(&addonsv1alpha1.HelmReleaseProxy{}).
 		WithEventFilter(predicate.GenerationChangedPredicate{}).
 		WithEventFilter(predicates.ResourceNotPausedAndHasFilterLabel(log, r.WatchFilterValue)).
+		Watches(
+			&clusterv1.Cluster{},
+			handler.EnqueueRequestsFromMapFunc(clusterToHelmReleaseProxies),
+			builder.WithPredicates(
+				predicates.All(ctrl.LoggerFrom(ctx),
+					predicates.Any(ctrl.LoggerFrom(ctx),
+						predicates.ClusterUnpaused(ctrl.LoggerFrom(ctx)),
+						predicates.ClusterControlPlaneInitialized(ctrl.LoggerFrom(ctx)),
```

Comment:
@CecileRobertMichon I added this predicate but need some help configuring the logic. The Cluster watch seems to be blocked by `WithEventFilter(predicate.GenerationChangedPredicate{})` on line 65, which only includes updates where the HelmReleaseProxy spec has changed. The intention of that filter was to fix a bug where it would requeue itself when the controller changes a field in the CRD, but this seems to also be filtering out the `predicates.ClusterControlPlaneInitialized`. Is there a way to have the `predicate.GenerationChangedPredicate` not apply to the Cluster watch too?

Reply:
> The intention of that filter was to fix a bug where it would requeue itself when the controller changes a field in the CRD

What field is the controller changing on the CRD? is there a way to fix the bug itself to avoid having to use that filter altogether?

Reply:
When `spec.ReleaseName` is empty, the HelmReleaseProxy controller will generate one and set it in a `Patch()` at the end of the reconciliation, and the same goes for the status fields like conditions. I worked on this a long time ago when I was prototyping, but I believe the intention was that we only wanted the controller requeue when there's actually something to change, i.e. the HelmChartProxy triggered an update.

Reply:
I can try to do some triaging, but in the case that it is an issue with `spec.ReleaseName`, I'm thinking we would have to redesign it by either moving it to the status or making it immutable and set on CRD creation.

Reply:
@CecileRobertMichon So I did some more investigation and I can confirm that when I use `patchHelper.PatchObject()`, it causes the controller to requeue immediately even if there's no change. I'm not sure if this is a bug, or if this is just the normal behavior we expect to see in CAPZ/CAPI.

Reply:
I tested this with CAPI Cluster controller and confirmed that each time we use the patchHelper, we trigger a requeue. When creating a Cluster, the controller requeues once when we add the finalizer and patch immediately and a second time when we call patch in the defer function. Since it seems like this behavior in CAAPH is consistent with the CAPI controllers, I think we can consider the requeue behavior as non-blocking for this PR. WDYT?

Reply:
> When spec.ReleaseName is empty, the HelmReleaseProxy controller will generate one

This sounds like a defaulting behavior, could it be done in a webhook instead?

Reply:
Sorry, I checked and it looks like we're passing the `GenerateName` flag into the Helm install which is returning a release name to us. But it seems like we could still just have it be done in the HelmChartProxy controller. Either way, the patch function requeues whether we're setting the release name or not, though.

Reply:
is the `GenerateName` logic exposed as a public func in the Helm library? if so, could we call it from a webhook instead of waiting for the install to default the name?

Reply:
>  Either way, the patch function requeues whether we're setting the release name or not, though.

It's okay if it requeues once. What we don't want is an infinite reconciliation loop.

Reply:
No, it's a field we set on the install client like `installClient.GenerateName = true`. And to your other point, it doesn't produce an infinite reconcile loop as the requeues caused by Patch() do not trigger additional requeues.